### DISCUSSION TITLE
How to reference private repo composite actions from within other such actions

### DISCUSSION THREAD

**Discussion by kristofdho** [actions/runner#2431] [Q&A]
We have a private repository set up to host private actions, structured like this:
```
action1
├── action.yml
action2
└── action.yml
```

This works fine as long as the actions do not rely on each other. To simplify the setup, I would like to add a third action, that simplifies the other 2, which would need to be called like this:

```yml
steps:
  - uses: ${{ github.action_path }}/../action3
```

However, `github.action_path` is not substituted, which is also explained here: https://github.com/orgs/community/discussions/27048
This was a similar issue as well: https://github.com/actions/runner/discussions/1419 with the key difference being that the initial action is in the same repository as the subworkflow, while now they are in different repositories so a `uses: ./` would reference the wrong repository.

I get that the substitution does not work because it needs to be able to verify the source of the action ahead of time, but in this case, the caller is in the same repository as the callee, so if the caller is already allowed, it will also allow the callee.

I also know that this can be worked around by adding a self-reference like `uses: org/current-repo@tag` but I want the same tag to be used as the caller instead of referencing to a preexisting tag.

Is there a different way the new action can be referenced without having to resort to self-referencing the repository?

**Comment by SirHegel** [comment]
This is now supported on GitHub.com with the self-repository `$/` reference. In the parent composite action:

```yaml
# action1/action.yml
runs:
  using: composite
  steps:
    - name: Call the sibling action
      uses: $/action3
```

`$/` resolves against the repository containing the running action, at the same commit as that action. Therefore, if a caller uses `org/current-repo/action1@v2` (or pins a SHA), `$/action3` resolves to `action3` from that exact resolved commit. It needs no checkout and must not have an `@ref` suffix.

GitHub documents the self-repository reference and specifically notes that it resolves against the repository of the file in which it appears, rather than the caller's workspace. That is what makes it reliable for action composition: [workflow syntax](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#example-using-an-action-in-the-same-repository-as-the-workflow-at-the-running-commit-recommended).

One limitation: the documentation says `$/` is not available on GitHub Enterprise Server. On GHES, the explicit `org/repo/action3@ref` approach remains necessary.

**Reply by bigunyak** [reply]
When will it be available in GHES though?

**Reply by SirHegel** [reply]
@bigunyak GitHub has not published a GHES release or date for `$/` yet.

The [July 30 launch announcement](https://github.blog/changelog/2026-07-30-reference-same-repository-actions-with-self-repository-syntax/) says the feature is available on GitHub.com and requires runner 2.336.0 or newer. The current [GHES 3.22 workflow syntax](https://docs.github.com/en/enterprise-server@3.22/actions/reference/workflows-and-actions/workflow-syntax#example-using-an-action-in-the-same-repository-as-the-workflow-at-the-running-commit-recommended) still explicitly says that `$/` is unavailable on GitHub Enterprise Server. Therefore, installing a newer self-hosted runner alone is not a documented way to enable it.

Until a GHES release note removes that limitation, the supported fallback remains an explicit `owner/repository/path@ref` (ideally a full commit SHA). I would watch the GHES release notes rather than infer a date from the runner version.

**Reply by jenseng** [reply]
Another approach on GHES is to sync [dynamic-uses](https://github.com/jenseng/dynamic-uses) to your installation, and then use it in your actions/workflows. The usage is pretty simple:

```yaml
- uses: jenseng/dynamic-uses@v1
  with:
    # now you can use expressions 🥳
    uses: my-org/my-action@${{ inputs.some_ref }}
    # the `with` needs to be converted to a string (YAML mapping or JSON object)
    with: |
      some-input: yolo
```

See the README for more examples and details
